Add lagrange points L4 and L5

snumlautoken opened the pull request:
### Description

This code adds the option to enable markings of L4 and L5 lagrange points between the current body and the solar object.
<!--- Please also include relevant motivation and context. -->
As points L1, L2 and L5 are all analogue to solar and antisolar points, I saw no reason to add them.

No additional dependencies are required.

Fixes #1111 

### Screenshots (if appropriate):

### Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] This change requires a documentation update

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
Honestly, I have just been eyeballing the feature in action and debugging to see that my math is correct. The math is not particularly hard, but I am not sure how this could be tested automatically. Let me know if you want me to add anything.
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

**Test Configuration**:
* Operating system: Arch-linux
* Graphics Card: Not really applicable

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (header file)
- [ ] I have updated the respective chapter in the Stellarium User Guide
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

alex-w commented:
OK, I see 2 Lagrange points for Earth-Sun system and this is good, but OP in the issue #1111 want to see L4 & L5 points for Earth-Moon system. 

To avoid misunderstanding I propose rename "L4" into "L4 (Sun-planet)" and "L5" into "L5 (Sun-planet)" in the GridLinesMgr.cpp file for current L-points. What about L-points for the Moon?

Atque commented:
> As points L1, L2 and L3 are all analogue to solar and antisolar points, I saw no reason to add them.

Wouldn't that be good though, for educational reasons? And to understand the numbering of the lagrange points.

snumlautoken commented:
@alex-w 
I have renamed the existing lagrange-points and added the L4 and L5 points for the earth-moon system

@Atque 
Thought about it, but wouldn't it be cluttered? Especially seeing as how L1 and L3 occupy the same point (from the observer).
Would be easy to add if requested though.

gzotti commented on src/core/modules/GridLinesMgr.cpp:
```diff
@@ -1756,7 +1777,35 @@ void SkyPoint::draw(StelCore *core) const
+		{
+			// Takes the positional vector of the planet from the solar object
+			Vec3d pos= core->getCurrentObserver()->getHomePlanet()->getHeliocentricEclipticPos();
+			// Creates points by adding/subtracting 60 degrees to the vector along the ecliptic plane
```
This approach may work for circular orbits. How does it appear for e.g. Mars or Mercury? Its distance from the sun will place it far off the orbit distance at that point.

snumlautoken replied:
See commit [78d81e7](https://github.com/Stellarium/stellarium/pull/3090/commits/78d81e7851306c086473ab1541b39cb84eb1b214) 
I am thinking it should account for deviations from orbit, but I am not really an astronomer.

alex-w commented:
Earth-Moon lagrange points should be visible from Earth only

gzotti commented on src/core/modules/GridLinesMgr.cpp:
```diff
@@ -3920,6 +3986,63 @@ void GridLinesMgr::setColorApexPoints(const Vec3f& newColor)
+		lagrangePointsSolar->setColor(newColor);
+		emit lagrangePointsSolarColorChanged(newColor);
+	}
+
```
remove empty line

gzotti commented:
L1, 2, 3 would be more clearly visible when observed from the associated "planet observer". Also, if they are computed with the planetocenter, an observer on the surface will see at least the distinction of L1 from L3 from parallax shift, even if only a few arcseconds. (But L3 is hidden behind the sun...)

snumlautoken commented:
> L1, 2, 3 would be more clearly visible when observed from the associated "planet observer". Also, if they are computed with the planetocenter, an observer on the surface will see at least the distinction of L1 from L3 from parallax shift, even if only a few arcseconds. (But L3 is hidden behind the sun...)

If that is the case, wouldn't we need the masses of the bodies to calculate the distances? As far as I can tell the stellar objects classes have no members for mass, so this would not work for the general case.
Please correct me if I am wrong.

alex-w reviewed (approved):
Well, we don't have masses of the SS bodies in our INI files, I think other LP should be implemented in separate PR if it critical. The current PR need couple cosmetic fixes, but it can be fixed after merge.

Atque commented:
I think the labels should be "Sun-< name of current planet >" instead of "Sun-Planet".

gzotti commented:
When observing from Earth Observer, I expect to find L4, L5 somewhere close to Earth's orbit line. However, they are at no location that appear to make sense.

snumlautoken commented:
> When observing from Earth Observer, I expect to find L4, L5 somewhere close to Earth's orbit line. However, they are at no location that appear to make sense.

Yes, I just awoke to the realization that I am calculating the common plane incorrectly, again...
It has been ages since I last touched linear algebra.
I will try to resolve it soon.

snumlautoken commented:
@gzotti 
If you would please have a look at [ee0966a](https://github.com/Stellarium/stellarium/pull/3090/commits/ee0966a3e141ab2a3e7dfa7f912718e964364ffa).
Been sure to make my testing more extensive and can at the very least confirm that the vectors are equidistant.
Sorry for all the trouble.

alex-w commented:
> When observing from Earth Observer, I expect to find L4, L5 somewhere close to Earth's orbit line. However, they are at no location that appear to make sense.

Wow, but this is requires add special implementation for the `observer` type, because currently all computations are connected to the current planet.

snumlautoken commented:
I may misunderstand the issue, but when I am observing from Earth Observer L4 and L5 are very close to the orbit line.

![image](https://user-images.githubusercontent.com/28626221/224383304-db580ba3-0324-49d1-874e-7b444fed416e.png)
![image](https://user-images.githubusercontent.com/28626221/224383602-54cda58e-acf2-4af8-b9a1-9d404ad5ecd6.png)

gzotti commented:
If you then move away from the earth (Alt+End), you will see this.
 
![stellarium-072](https://user-images.githubusercontent.com/18099873/224439855-69a6fd3e-4f48-473f-b32c-2fd343715049.jpg)

So, something is still way off. But I have not looked into the code or searched for actual recommended algorithms.

alex-w commented:
> If you then move away from the earth (Alt+End), you will see this.
> 
> ![stellarium-072](https://user-images.githubusercontent.com/18099873/224439855-69a6fd3e-4f48-473f-b32c-2fd343715049.jpg)
> 
> So, something is still way off. But I have not looked into the code or searched for actual recommended algorithms.

Hmm… I see observercentric computations.

alex-w reviewed (changes requested):
OK, some cosmetic fixes (like: “L4(Sun-planet)” -> “L4 (Sun-planet)”) may be fixes before merging.

snumlautoken commented:
> If you then move away from the earth (Alt+End), you will see this.

Gotcha! This has been fixed!
![image](https://user-images.githubusercontent.com/28626221/224477550-ad56e6fe-e2b4-4a37-8e4c-6b93315a3d79.png)

snumlautoken commented:
> OK, some cosmetic fixes (like: “L4(Sun-planet)” -> “L4 (Sun-planet)”) may be fixes before merging.

Also fixed this in the commit, though it is not visible in the screenshot as I had not added it at the time that the screenshot was taken.

gzotti commented:
OK, seems we are getting somewhere. Now in this view, L1,2,3 would be a welcome addition. So, maybe show L1,2,3 if observed from an observer location? 

The other thing is the label. IMO it is too much screen clutter. "L`<n>` (`<planetname>`)" or even "L`<n>`<sub>`<planetsymbol>`</sub>" should be enough. 

The next question is of course the actual reason for showing L points. OP wanted to see the location of the elusive Kordylewski clouds in the Moon's L4,5. AS soon as we have that, other people will become interested in seeing L4,5 of the other planets in search of what are Trojans for Jupiter. This would mean, 8*2 extra points (or 8*5 if observed from an "observer") over the whole sky if they are plotted for every planet. Or a fine-grained selection of which planet's L points the users are interested.

snumlautoken commented:
> Now in this view, L1,2,3 would be a welcome addition

Is there a way to get my hands on the mass of the planet without hard-coding it? As far as I can tell it will be necessary for the calculation of the points.

> This would mean, 82 extra points (or 85 if observed from an "observer") over the whole sky if they are plotted for every planet

I could imagine adding a drop-down menu for other planets in the SS, so while the current planet would be the default the user would have the option to choose whatever other planet in the SS. Does that sound reasonable?

gzotti commented:
We could introduce an optional mass_kg entry (default 0) to the ssystem_*.ini files, massKg instance variable and getMass() function to the Planet class.

alex-w commented:
> OK, seems we are getting somewhere. Now in this view, L1,2,3 would be a welcome addition. So, maybe show L1,2,3 if observed from an observer location?
> 
> The other thing is the label. IMO it is too much screen clutter. "L`<n>` (`<planetname>`)" or even "L`<n>``<planetsymbol>`" should be enough.
> 
> The next question is of course the actual reason for showing L points. OP wanted to see the location of the elusive Kordylewski clouds in the Moon's L4,5. AS soon as we have that, other people will become interested in seeing L4,5 of the other planets in search of what are Trojans for Jupiter. This would mean, 8x2 extra points (or 8x5 if observed from an "observer") over the whole sky if they are plotted for every planet. Or a fine-grained selection of which planet's L points the users are interested.

Showing L-points for selected planet or for current planet when any planet is not selected?

gzotti commented:
> > The other thing is the label. IMO it is too much screen clutter. "L`<n>` (`<planetname>`)" or even "L`<n>`<sub>`<planetsymbol> `</sub>" should be enough.
> 
> Bad idea for performance

What is so costly here? String creation in SkyPoint::updateLabel() is done once whenever language changes. String display? Is a short but complex string so much slower to draw than a long one? How much? If the sub is costly, we can omit it. But short labels should in any case be used to avoid screen clutter.

> > The next question is of course the actual reason for showing L points. OP wanted to see the location of the elusive Kordylewski clouds in the Moon's L4,5. AS soon as we have that, other people will become interested in seeing L4,5 of the other planets in search of what are Trojans for Jupiter. This would mean, 8x2 extra points (or 8x5 if observed from an "observer") over the whole sky if they are plotted for every planet. Or a fine-grained selection of which planet's L points the users are interested.
> 
> Showing L-points for selected planet or for current planet when any planet is not selected?

A choice of up to 8 planets plus Earth's Moon for which L points could be made available. This would call for a new sub-dialog available from the Gridline viewconfig page. It should be possible to display L4,5 for Jupiter and Saturn while having selection on some other object.

However, this all comes not in time for the upcoming release. Please let us discuss this further after 23.1 is out, there are still tasks to do.

alex-w commented:
> What is so costly here? String creation in SkyPoint::updateLabel() is done once whenever language changes. String display? Is a short but complex string so much slower to draw than a long one? How much? If the sub is costly, we can omit it. But short labels should in any case be used to avoid screen clutter.

You need call SolarSystem/Planet classes every frame to get the localised name of the planets.

gzotti commented:
> > What is so costly here? String creation in SkyPoint::updateLabel() is done once whenever language changes. String display? Is a short but complex string so much slower to draw than a long one? How much? If the sub is costly, we can omit it. But short labels should in any case be used to avoid screen clutter.
> 
> You need call SolarSystem/Planet classes every frame to get the localised name of the planets.

updateLabels() is called when language changes, not in every frame. This of course assumes there are L points for each planet. If there is only a set of max. 5 L points for the current planet (or observed planet if user is on an observer-type planet), labels have to be recreated also if observed planet changes.

snumlautoken commented:
Things currently seem to be in limbo, so what exactly should we do to move forward with the PR?

gzotti commented:
Sorry, we are busy with real-world jobs and bugfixes for 23.1. 

I wonder whether it makes sense to have up to all 8x5 L points for all planets and maybe even some moons of the gas giants which have other moons in their L4/5 points visualized in SolarSystemObserver view. This would require a dedicated selection GUI. 

If you need planet mass, you could find and insert mass data for the planets in ssystem_major.ini. Just add lines like 

`mass_kg =  5.972168e24
`

snumlautoken commented:
I was thinking about just adding a dropdown-menu to select the body whose lagrange-points to show; or alternatively to show those of the current cursor-selected body. Wouldn't that suffice?
Thinking it would be a little wonky with a dediccated selection GUI.
![image](https://user-images.githubusercontent.com/28626221/225081216-8885e534-b660-40eb-81dd-a0311f97a52f.png)

snumlautoken commented:
I also figure that I might as well map the constant distances of L1-3 instead of using their masses if it is not going to be a general case either way.

alex-w commented:
Masses of the Sun and major planets in the master already

alex-w commented:
Any news?

snumlautoken commented:
> Any news?

Ah, sorry, got busy with other work, so no progress has been made since
If I find time for it I could get it done for 23.2

gzotti commented:
@snumlautoken Any updates here?

alex-w commented:
@nimlapong why the pull request is closed?
